internal/alb/tg/targethealth.go

```diff
@@ -0,0 +1,358 @@
+	in := &elbv2.DescribeTargetHealthInput{TargetGroupArn: aws.String(tgArn)}
+	resp, err := c.cloud.DescribeTargetHealthWithContext(ctx, in)
+	if err != nil {
+		return err
+	}
+	targetsHealth := map[string]*elbv2.TargetHealth{}
+	for _, desc := range resp.TargetHealthDescriptions {
+		targetsHealth[*desc.Target.Id] = desc.TargetHealth
+	}
+
+	pods, err := c.endpointResolver.ReverseResolve(ingress, backend, targetsToReconcile)
+	if err != nil {
+		return err
+	}
+
+	for _, pod := range pods {
+		targetHealth, ok := targetsHealth[pod.Status.PodIP]
+		if ok && podHasReadinessGate(pod, conditionType) {
+			if err := c.reconcilePodCondition(ctx, conditionType, pod, targetHealth, true); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// Creates or updates the condition status for the given pod with the given target health.
+func (c *targetHealthController) reconcilePodCondition(ctx context.Context, conditionType api.PodConditionType, pod *api.Pod, targetHealth *elbv2.TargetHealth, updateTimes bool) error {
+	conditionStatus := podConditionStatusFromTargetHealth(targetHealth)
+
+	// check if condition already exists
+	now := metav1.Now()
+	i, cond := podConditionForReadinessGate(pod, conditionType)
+	if cond == nil {
+		// new condition
+		targetHealthCondition := api.PodCondition{
+			Type:   conditionType,
+			Status: conditionStatus,
+			Reason: aws.StringValue(targetHealth.Reason),
+		}
+		if updateTimes {
+			targetHealthCondition.LastProbeTime = now
+			targetHealthCondition.LastTransitionTime = now
+		}
+		pod.Status.Conditions = append(pod.Status.Conditions, targetHealthCondition)
+	} else {
+		// update condition
+		if updateTimes {
+			cond.LastProbeTime = now
+			if cond.Status != conditionStatus {
+				cond.LastTransitionTime = now
+			}
+		}
+		cond.Status = conditionStatus
+		cond.Reason = aws.StringValue(targetHealth.Reason)
+		pod.Status.Conditions[i] = *cond
+	}
+
+	// pod will always be updated (at least to update `LastProbeTime`);
+	// this will trigger another invocation of `Reconcile`, which will remove this pod from the list of pods to reconcile if its health status is ok
+	err := c.client.Status().Update(ctx, pod)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// From the given targets, only returns the ones that have a readiness gate for the given ingress/service and whose pod conditions actually need to be reconciled.
+func (c *targetHealthController) filterTargetsNeedingReconciliation(conditionType api.PodConditionType, t *Targets, desiredTargets []*elbv2.TargetDescription) ([]*elbv2.TargetDescription, error) {
+	targetsToReconcile := []*elbv2.TargetDescription{}
+	if len(desiredTargets) == 0 {
+		return targetsToReconcile, nil
+	}
+
+	strategy, err := ingressTargetHealthReconciliationStrategy(t.Ingress)
```

Comment:
if we really want to support a reconcile strategy(to expose pod's status to user?), it should be derived from Ingress & Service, while annotation on services takes priority. since it's an attribute related to targetGroups.